### ISSUE TITLE
Pango install fail on 10.11

### ISSUE THREAD

**Issue by mikestecker** [Homebrew/homebrew#41086]
I know I'm running a beta, but wanted to report this anyway.

```
==> Installing fontforge dependency: pango
==> Downloading https://download.gnome.org/sources/pango/1.36/pango-1.36.8.tar.xz
==> Downloading from https://mirror.umd.edu/gnome/sources/pango/1.36/pango-1.36.8.tar.xz
######################################################################## 100.0%
==> ./configure --disable-silent-rules --prefix=/usr/local/Cellar/pango/1.36.8_1 --enable-man --with-html-dir=/usr/local/Cellar/pango/1.36.8_1/share/doc --enable-introspec
==> make
make[4]: *** [PangoCairo-1.0.gir] Error 1
make[3]: *** [all-recursive] Error 1
make[2]: *** [all] Error 2
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

I have the latest Xcode beta (beta 2) installed:
```
xcode-select -p
/Applications/Xcode-beta.app/Contents/Developer

xcode-select -v
xcode-select version 2342
```

Logs: https://gist.github.com/98ae04d68f1c8fa409df

**Comment by tdsmith** [comment]
It looks like g-ir-scanner is choking on the system Carbon headers: https://gist.github.com/anonymous/98ae04d68f1c8fa409df#file-02-make-L322-L351

**Comment by mmindenhall** [comment]
Just saw the same error with current released version of Xcode:
```
$ xcode-select -v
xcode-select version 2339.
```

**Comment by mmindenhall** [comment]
This worked for me:

1.  `brew info pango`
1.  `brew uninstall --force <all listed dependencies>`
1.  `brew install pango`

**Comment by DomT4** [comment]
```
Build: xz ✔, pkg-config ✔
Required: glib ✔, cairo ✔, harfbuzz ✔, fontconfig ✔, gobject-introspection ✔
```

Solution is probably to rebuild `gobject-introspection` then, I'd imagine. If that works as a solution for @mikestecker as well I'll stick it in the 10.11 tracking list as a workaround.

**Comment by mikestecker** [comment]
At first I tried this based off @DomT4's suggestion:

```
brew uninstall --force object-introspection
brew install pango
```

But this still failed to install with the same errors as I originally stated.

Then I tried to uninstall the remaining dependancies listed under "Required": `brew uninstall --force fontconfig harfbuzz cairo glib` & then `brew install pango`. Still no luck. Dependancies installed fine, but not pango.

I then tried uninstall the remaining 2 dependancies: `xz` & `pig-config` and then re-trying to install pango, but none of this worked still.

Lastly, I tried to do the exact steps that @mmindenhall did:

1. `brew info pango`:
```
pango: stable 1.36.8, HEAD
Framework for layout and rendering of i18n text
http://www.pango.org/
Not installed
From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/pango.rb
==> Dependencies
Build: xz ✔, pkg-config ✔
Required: glib ✔, cairo ✔, harfbuzz ✔, fontconfig ✔, gobject-introspection ✔
==> Options
--universal
	Build a universal binary
--HEAD
	Install HEAD version
```

2. `brew uninstall --force xz pkg-config glib cairo harfbuzz fontconfig gobject-introspection`
```
Uninstalling xz... (59 files, 1.6M)
Uninstalling pkg-config... (10 files, 600K)
Uninstalling glib... (424 files, 18M)
Uninstalling cairo... (111 files, 6.0M)
Uninstalling harfbuzz... (79 files, 3.4M)
Uninstalling fontconfig... (448 files, 3.6M)
Uninstalling gobject-introspection... (224 files, 11M)
```

3. `brew install pango`
```
Warning: You are using OS X 10.11.
We do not provide support for this pre-release version.
You may encounter build failures or other breakage.
==> Installing dependencies for pango: xz, pkg-config, glib, fontconfig, cairo, gobject-introspection, harfbuzz
```
[continued installing...]
```
==> Installing pango
==> Downloading https://download.gnome.org/sources/pango/1.36/pango-1.36.8.tar.xz
Already downloaded: /Library/Caches/Homebrew/pango-1.36.8.tar.xz
==> ./configure --disable-silent-rules --prefix=/usr/local/Cellar/pango/1.36.8_1 --enable-man --with-html-dir=/usr/local/Cellar/pango/1.36.8_1/share/doc --enable-introspec
==> make
make[4]: *** [PangoCairo-1.0.gir] Error 1
make[3]: *** [all-recursive] Error 1
make[2]: *** [all] Error 2
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

So that didn't fix it for me. Not sure what else, I have the latest Xcode Beta 2

![screenshot 2015-06-26 08 16 46](https://cloud.githubusercontent.com/assets/407465/8380631/bb63ae7c-1bdb-11e5-9eff-a76de0b52f70.png)

and it's set to be my CLT:
```
xcode-select -p
/Applications/Xcode-beta.app/Contents/Developer
```
![screenshot 2015-06-26 08 18 15](https://cloud.githubusercontent.com/assets/407465/8380682/eb20c97e-1bdb-11e5-93c4-3fad9f5cdc39.png)

Here are my logs after the last failure I had: https://gist.github.com/3babf899910d18c1ef3e

**Comment by mikeakers** [comment]
It's an ugly workaround, but I was able to get pango installed by commenting out lines 17 and 298 of `/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/CoreGraphics.framework/Headers/CGDisplayStream.h`

These macros have to do with Swift nullability annotations, so I doubt that disabling them will affect pango.

**Comment by tdsmith** [comment]
Thanks for the pointer. Can someone surface this to the gobject-introspection devs and link to the ticket here?

**Comment by mikestecker** [comment]
@mikeakers that workaround isn't working for me

**Comment by printfn** [comment]
I can confirm that @mikeakers workaround works.

**Comment by mikestecker** [comment]
@printfn I'm not sure what I'm missing then? I first tried commenting them out, then even tried simply deleting the two lines... still not working.

I'm supposed to be commenting out `CF_ASSUME_NONNULL_BEGIN` (17) and `CF_ASSUME_NONNULL_END` (298) correct?

here's my log: https://gist.github.com/8f691e91ee7613344e85

**Comment by printfn** [comment]
You are on 10.11 beta 2, with Xcode 7 beta 2 installed and selected using `xcode-select`, right?
I only needed to comment out lines 17 and 298 for the pango install to work.
I also reinstalled all of pango's dependencies, but I'm not sure if that was necessary.
Sorry, but I'm really not sure where the problem is.

**Comment by mikestecker** [comment]
Wow, just checked and `xcode-select` was somehow changed back to the regular version of Xcode... I just manually changed it using `sudo xcode-select -s /Applications/Xcode-beta.app/Contents/Developer`

But still no luck. I just removed all dependancies again using `brew uninstall --force xz pkg-config glib cairo harfbuzz fontconfig gobject-introspection` and tried reinstalling with those two lines from the beta version commented out but it's still not working.

Yes, I have 10.11 beta 2 and Xcode 7 beta 2 installed.

![screenshot 2015-06-26 16 18 41](https://cloud.githubusercontent.com/assets/407465/8389034/10fd50e8-1c1f-11e5-8abb-c92c0a78eb4b.png)
![screenshot 2015-06-26 16 18 53](https://cloud.githubusercontent.com/assets/407465/8389035/1116505c-1c1f-11e5-898d-0b35544023c5.png)

**Comment by mikestecker** [comment]
Logs: https://gist.github.com/d440bded2c6f97edc5e7

**Comment by mikestecker** [comment]
It appears after inspecting my logs here https://gist.github.com/anonymous/d440bded2c6f97edc5e7#file-01-configure-cc

![screenshot 2015-06-26 16 29 21](https://cloud.githubusercontent.com/assets/407465/8389141/a7a0cc5e-1c20-11e5-806c-e49387748dad.png)

that my system is not using the Xcode beta version? How can I switch it?

**Comment by mikestecker** [comment]
`brew config` spits out:

```
HOMEBREW_VERSION: 0.9.5
ORIGIN: https://github.com/Homebrew/homebrew
HEAD: db006687ea7b33e87c1131712b072258819febd1
Last commit: 3 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CELLAR: /usr/local/Cellar
CPU: 8-core 64-bit haswell
OS X: 10.11-x86_64
Xcode: 7.0 => /Applications/Xcode-beta.app/Contents/Developer
CLT: 7.0.0.0.1.1434774289
Clang: 7.0 build 700
X11: 2.7.7 => /opt/X11
System Ruby: 2.0.0-p645
Perl: /usr/bin/perl
Python: /usr/local/bin/python => /usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/bin/python2.7
Ruby: /Users/mikestecker/.rbenv/shims/ruby => /Users/mikestecker/.rbenv/shims/ruby
Java: N/A
```

**Comment by printfn** [comment]
Do you have the environment variable [DEVELOPER_DIR](http://stackoverflow.com/a/18682905) set? It might override xcode-select. Otherwise I'm pretty much out of ideas.

**Comment by mikestecker** [comment]
@printfn I don't see anything with that in it. I tried Googling how to find where it might be set or how to clear it but can't seem to find anything.

**Comment by asyncsrc** [comment]
Hi @mikestecker -- I have the exact same versions as you of MacOS, Xcode, etc., and I ended up resolving the issue by only commenting out one line from here */System/Library/Frameworks/CoreGraphics.framework/Headers/CGDisplayStream.h*:

Line 15: `//CF_IMPLICIT_BRIDGING_ENABLED`

It seems like after commenting out that single line, pango now installs properly via homebrew.

```
╰─ brew info pango
pango: stable 1.36.8, HEAD
Framework for layout and rendering of i18n text
http://www.pango.org/
/usr/local/Cellar/pango/1.36.8_1 (124 files, 4.6M) *
  Built from source
From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/pango.rb
```

```
╰─ xcode-select -v
xcode-select version 2342.
```

```
╰─ xcode-select -p
/Applications/Xcode-beta.app/Contents/Developer
```

Hopefully that will be of some help

**Comment by mikestecker** [comment]
@asyncsrc how did you edit the file? OS X seems to be still locking me out. I tried using `sudo` but I'm still getting `Operation not permitted` from nano to edit the file.

**Comment by asyncsrc** [comment]
@mikestecker honestly I just did `sudo vim /System/Library/Frameworks/CoreGraphics.framework/Headers/CGDisplayStream.h`

However, I'm not sure if it's related, but I recently found out about 10.11 having a new feature called "rootless" where even with root, there are some folders that I was unable to modify.  It's possible that /System/Library/* is included, but I'm not entirely sure.

Referenced here -- https://github.com/haskell/cabal/issues/2653, one of the guys suggested disabling rootless mode by doing -- `sudo nvram boot-args="rootless=0"` at the commandline then rebooting.

That's about the only reason I can imagine that you'd be unable to edit that file, even when using sudo.

**Comment by DomT4** [comment]
You won't be able to touch (almost) anything in `/System` with `rootless`. It's a deliberate choice by Apple. If you wanted to edit the header you'd need to do so inside the `Xcode` SDK rather than inside `/System` itself.

FWIW, I don't recommend disabling rootless.

**Comment by asyncsrc** [comment]
Right, if you'd like you can disable rootless, update the CGDisplayStream.h, and then enable rootless again with `sudo nvram boot-args="rootless=1` then reboot to re-enable it.

Or there's probably an alternate route that DomT4 suggests, but I don't have experience there.

Thank you @DomT4 for the info regarding rootless and /system.

**Comment by tdsmith** [comment]
Has anyone filed an issue against gobject-introspection yet?

**Comment by gabrielecirulli** [comment]
Is there any solution to this? I've tried all of the solutions discussed but with no luck.

**Comment by mikestecker** [comment]
Pango now installs after updating to El Capitan Beta 3 and the latest CMT through Software Update!

**Comment by DomT4** [comment]
Can confirm the above. Forgot how long it takes to install all the Pango dependencies from source, but the latest beta/xcode/CLT/etc seem to have resolved the issue. Will close as fixed later today unless anyone still has breakage.

**Comment by felixbuenemann** [comment]
This issue only affects 10.11 Beta 2 and Xcode 7 Beta 2 due to wrong nesting of preprocessor macros in `CGDisplayStream.h`. It's resolved in 10.11 Beta 3 and Xcode 7 Beta 3, I've checked the headers in Xcode and `/System/Library/Frameworks/CoreGraphics.framework`.

If anyone is stuck on the old versions this is the patch I proposed to Apple on June 24th:
```diff
--- /System/Library/Frameworks/CoreGraphics.framework/Versions/A/Headers/CGDisplayStream.h	2015-06-18 06:41:31.000000000 +0200
+++ /System/Library/Frameworks/CoreGraphics.framework/Versions/A/Headers/CGDisplayStream.h	2015-06-24 04:16:27.000000000 +0200
@@ -295,10 +295,10 @@
 CG_EXTERN CFRunLoopSourceRef __nullable CGDisplayStreamGetRunLoopSource(CGDisplayStreamRef __nullable displayStream)
     CG_AVAILABLE_STARTING(__MAC_10_8,__IPHONE_NA);
 
+#endif /* __BLOCKS__ */
+
 CF_ASSUME_NONNULL_END
 
 CF_IMPLICIT_BRIDGING_DISABLED
 
-#endif /* __BLOCKS__ */
-
 #endif /* CGDISPLAYSTREAM_H_ */
```

The changes can be applied by disabling rootless or from recovery mode using vim.
